browse/services/database/__init__.py

```diff
@@ -155,3 +205,147 @@ def get_dblp_authors(paper_id: str) -> List[str]:
         order_by(DBLPDocumentAuthor.position).all()
     authors = [a for (a,) in authors_t]
     return authors
+
+
+@db_handle_error(logger=logger, default_return_val=None)
+def get_document_count() -> Optional[int]:
+    """Get the number of documents."""
+    # func.count is used here because .count() forces a subquery which
+    # is inefficient
+    row = db.session.query(
+            func.count(Document.document_id).label('num_documents')
+          ).filter(not_(Document.paper_id.like('test%'))).first()
+    return row.num_documents
+
+
+@db_handle_error(logger=logger, default_return_val=None)
+def get_sequential_id(paper_id: Identifier,
+                      context: str = 'all',
+                      is_next: bool = True) -> Optional[str]:
+    """Get the next or previous paper ID in sequence."""
+    local_session = db.session()
+    baked_query = bakery(lambda session: session.query(Document.paper_id))
+
+    if paper_id.is_old_id:
+        # NB: classic did not support old identifiers in prevnext
+        if context == 'all':
+            like_id = f'{paper_id.archive}/{paper_id.yymm}%'
+        else:
+            like_id = f'%/{paper_id.yymm}%'
+    else:
+        like_id = f'{paper_id.yymm}.%'
+
+    baked_query += lambda q: q.filter(
```

Comment:
I realize that the SQLAlchemy docs shows examples with lambdas like this, too, but it sure smells a lot like assigning a lambda (see https://www.python.org/dev/peps/pep-0008/#programming-recommendations)... any reason not to explicitly define these?

Reply:
The `BakedQuery` extension is the root of this construction: https://docs.sqlalchemy.org/en/13/orm/extensions/baked.html

The primary motive for using this system is for performance increases when composing a query. It would be possible to edit the `Query` object iteratively like so:
```
query = Session.query(Document.paper_id)
query = query.filter(Document.paper_id.like(bindparam('like_id'))
```

However, by using the `baked()` construction, we get the added benefit of a query cache, speeding up subsequent calls to the query.

One question that remains though is whether, because the queries in each of these functions is executed once per request, whether we're introducing overhead for no benefit at all -- if a new baked query is getting constructed on each function call to `services.database`, then the cache isn't doing anything and clarity of code might be the bigger advantage.

Reply:
> One question that remains though is whether, because the queries in each of these functions is executed once per request, whether we're introducing overhead for no benefit at all -- if a new baked query is getting constructed on each function call to `services.database`, then the cache isn't doing anything and clarity of code might be the bigger advantage.

Actually, that's not quite right (although it's certainly what one would expect given typical Python scoping rules).  From the docs:

> In the above code, even though our application may call upon search_for_user() many times, and even though within each invocation we build up an entirely new BakedQuery object, all of the lambdas are only called once. Each lambda is never called a second time for as long as this query is cached in the bakery.

> The caching is achieved by storing references to the lambda objects themselves in order to formulate a cache key;

In other words, there is a singleton behind the scenes somewhere that is storing a reference to the lambdas. Which...is pretty much what PEP8 is advising against.

Reply:
Here is an untested refactor of this function without the bakery that reduces the number of lines by ~18%.

```python
@db_handle_error(logger=logger, default_return_val=None) 
def get_sequential_id(paper_id: Identifier,
                      context: str = 'all',
                      is_next: bool = True) -> Optional[str]:
    """Get the next or previous paper ID in sequence."""
    local_session = db.session()
    if paper_id.is_old_id:
        # NB: classic did not support old identifiers in prevnext
        if context == 'all':
            like_id = f'{paper_id.archive}/{paper_id.yymm}%'
        else:
            like_id = f'%/{paper_id.yymm}%'
    else:
        like_id = f'{paper_id.yymm}.%'

    q = local_session.query(Document.paper_id) \
        .filter(Document.paper_id.like(like_id))

    if is_next:
        q = q.filter(Document.paper_id > paper_id) \
            .order_by(asc(Document.paper_id))
    else:
        q = q.filter(Document.paper_id < paper_id) \
            .order_by(desc(Document.paper_id))

    if context != 'all':
        archive: str = context
        subject_class: str = ''
        if '.' in context:
            archive, subject_class = archive.split('.', 1)
        q = q.join(in_category).filter(in_category.c.archive == archive)
        if subject_class:
            q = q.filter(in_category.c.subject_class == subject_class)

    result = q.first()
    if result:
        return f'{result.paper_id}'
    return None
```

Reply:
https://github.com/arXiv/arxiv-browse/pull/84

Reply:
Thanks for the summary of our discussion @erickpeirson. These are all good reasons to move away from baked queries, and refactoring is not a big deal. I'll resolve this conversation for now. Requested changes are reflected in #84.